addons/l10n_vn_edi/models/account_move.py

```diff
@@ -0,0 +1,719 @@
+
+        # if the above request did not return data, we can assume that the invoice has failed to be created, or was never sent
+        if not invoice_data:
+            # Send the invoice to the system
+            request_response, error_message = self._l10n_vn_edi_send_request(
+                method='POST',
+                endpoint=f'InvoiceAPI/InvoiceWS/createInvoice/{self.company_id.vat}',
+                json_data=invoice_json_data,
+            )
+
+            if error_message:
+                return [error_message]
+
+            invoice_data = request_response['result']
+
+        self.write({
+            'l10n_vn_edi_reservation_code': invoice_data['reservationCode'],
+            'l10n_vn_edi_invoice_number': invoice_data['invoiceNo'],
+            'l10n_vn_edi_state': 'sent',
+        })
+
+    def _l10n_vn_edi_cancel_invoice(self, reason, agreement_document_name, agreement_document_date):
+        """ Send a request to cancel the invoice. """
+        self.ensure_one()
+        _request_response, error_message = self._l10n_vn_edi_send_request(
+            method='POST',
+            endpoint='InvoiceAPI/InvoiceWS/cancelTransactionInvoice',
+            params={
+                'supplierTaxCode': self.company_id.vat,
+                'templateCode': self.l10n_vn_edi_invoice_symbol.invoice_template_id.name,
+                'invoiceNo': self.l10n_vn_edi_invoice_number,
+                'strIssueDate': self._l10n_vn_edi_format_date(self.l10n_vn_edi_issue_date),
+                'additionalReferenceDesc': agreement_document_name,
+                'additionalReferenceDate': self._l10n_vn_edi_format_date(agreement_document_date),
+                'reasonDelete': reason,
+            },
+        )
+        if error_message:
+            raise UserError(error_message)
+
+        self.l10n_vn_edi_state = 'cancelled'
+        self.with_context(no_new_invoice=True).message_post(
+            body=_('The invoice has been cancelled for reason: %s', reason),
+        )
+        self.button_cancel()
+
+    def button_draft(self):
+        # EXTEND account
+        # When going from cancelled => draft, we ensure to clear the edi fields so that the invoice can be resent if required.
+        sinvoices = self.filtered(
+            lambda i: i.company_id.country_id.code == 'VN' and i.state == 'cancel'
+        )
+        sinvoices.write({
+            'l10n_vn_edi_transaction_id': False,
+            'l10n_vn_edi_invoice_number': False,
+            'l10n_vn_edi_reservation_code': False,
+            'l10n_vn_edi_issue_date': False,
+            'l10n_vn_edi_state': False,
+        })
+        # Cleanup the files as well. They will still be available in the chatter.
+        sinvoices.l10n_vn_edi_sinvoice_xml_file_id.unlink()
+        sinvoices.l10n_vn_edi_sinvoice_pdf_file_id.unlink()
+        sinvoices.l10n_vn_edi_sinvoice_file_id.unlink()
+        return super().button_draft()
```

Comment:
Make the super call first. If you get an error, no need to do all the stuff above.